src/python/pants/backend/python/framework/django/scripts/django_dependency_visitor.py

```diff
@@ -0,0 +1,48 @@
+# Copyright 2023 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+# -*- coding: utf-8 -*-
+
+# NB: This must be compatible with Python 2.7 and 3.5+.
+# NB: An easy way to debug this is to invoke it directly on a file.
+#   E.g.
+#   $ PYTHONPATH=src/python VISITOR_CLASSNAMES=pants.backend.python.framework.django.scripts\
+#     .django_dependency_visitor.DjangoDependencyVisitor \
+#     python src/python/pants/backend/python/dependency_inference/scripts/main.py FILE
+
+from __future__ import print_function, unicode_literals
+
+import ast
+
+from pants.backend.python.dependency_inference.scripts.dependency_visitor_base import (
+    DependencyVisitorBase,
+)
+
+
+class DjangoDependencyVisitor(DependencyVisitorBase):
+    def __init__(self, *args, **kwargs):
+        super(DjangoDependencyVisitor, self).__init__(*args, **kwargs)
+        self._in_migration = False
+
+    def visit_ClassDef(self, node):
+        # Detect `class Migration(migrations.Migration):`
+        if (
+            node.name == "Migration"
+            and len(node.bases) > 0
+            and node.bases[0].value.id == "migrations"
+            and node.bases[0].attr == "Migration"
+        ):
+            self._in_migration = True
+        self.generic_visit(node)
+        self._in_migration = False
+
+    def visit_Assign(self, node):
+        if self._in_migration and len(node.targets) > 0 and node.targets[0].id == "dependencies":
+            if isinstance(node.value, (ast.Tuple, ast.List)):
+                for elt in node.value.elts:
+                    if isinstance(elt, (ast.Tuple, ast.List)) and len(elt.elts) == 2:
+                        app = self.maybe_str(elt.elts[0])
+                        migration = self.maybe_str(elt.elts[1])
+                        if app is not None and migration is not None:
+                            module = "{}.migrations.{}".format(app, migration)
```

Comment:
This assumes that all apps are at a source root, which may not be the case.

Example (from a real project):

```
PANTS_SOURCE=../../kaos/pants pants dependencies src/python/org/proj/app1/migrations/0050_auto_20230117_1052.py
10:04:27.73 [INFO] View on BuildSense: https://app.toolchain.com/organizations/redacted/
10:04:29.40 [INFO] Parsed src/python/org/proj/app1/migrations/0050_auto_20230117_1052.py: {'imports': {'app2.migrations.0013_redacted': {'lineno': 9, 'weak': True}, 'app1.migrations.0049_redacted': {'lineno': 9, 'weak': True}, 'django.db.migrations': {'lineno': 3, 'weak': False}, 'django.db.models': {'lineno': 3, 'weak': False}, 'django.db.models.deletion': {'lineno': 4, 'weak': False}}, 'assets': []}
3rdparty/python#Django
```

Roots are:
```
build/scripts
src/python
tests/python
```

Reply:
I applied this diff to your PR in order to try it live:
```diff
diff --git a/src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py b/src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py
index e317121dc..f7165cf75 100644
--- a/src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py
+++ b/src/python/pants/backend/python/dependency_inference/parse_python_dependencies.py
@@ -1,6 +1,7 @@
 # Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
 # Licensed under the Apache License, Version 2.0 (see LICENSE).

+import logging
 import json
 import os
 from dataclasses import dataclass
@@ -202,6 +203,8 @@ async def parse_python_dependencies(
     process_output = process_result.stdout.decode("utf8") or "{}"
     output = json.loads(process_output)

+    logging.info(f"Parsed {request.source.address}: {output}")
+
     return ParsedPythonDependencies(
         imports=ParsedPythonImports(
             (key, ParsedPythonImportInfo(**val)) for key, val in output.get("imports", {}).items()
diff --git a/src/python/pants/backend/python/register.py b/src/python/pants/backend/python/register.py
index 36fd13714..ee69d506d 100644
--- a/src/python/pants/backend/python/register.py
+++ b/src/python/pants/backend/python/register.py
@@ -54,6 +54,9 @@ from pants.build_graph.build_file_aliases import BuildFileAliases
 from pants.core.target_types import TargetGeneratorSourcesHelperTarget
 from pants.core.util_rules.wrap_source import wrap_source_rule_and_target

+from pants.backend.python.framework.django import rules as django
+
+
 wrap_python = wrap_source_rule_and_target(PythonSourceField, "python_sources")


@@ -63,6 +66,7 @@ def build_file_aliases():

 def rules():
     return (
+        *django.rules(),
         *target_types_rules.rules(),
         # Subsystems
         *coverage_py.rules(),
```

Reply:
Hmm, there is a problem, but it's not with source roots. I was absent-mindedly assuming that the migration deps use the dotted module path (e.g., `django.contrib.auth`), which is from the source root, so the source root doesn't matter.

But actually it's the Django app "label", which is typically just the last component of the module path, e.g., `auth` but I think can be overridden in the `AppConfig` class (I will experiment). 

So it's not the source roots that are the issue, but that we need a map from app label to app module.

Reply:
Don't forget that this extraction phase returns dependency modules, not files, so it does not know or need to know about source roots.

Reply:
> But actually it's the Django app "label", which is typically just the last component of the module path, e.g., auth but I think can be overridden in the AppConfig class (I will experiment).

Ah, true.

> Don't forget that this extraction phase returns dependency modules, not files, so it does not know or need to know about source roots.

Well, I guess it doesn't need to know about source roots or file paths, but the dependency module won't work if it isn't complete from a source root perspective.

With a app label to module mapping this will work just fine, so I think that is the correct way forward :)

Reply:
> Well, I guess it doesn't need to know about source roots or file paths, but the dependency module won't work if it isn't complete from a source root perspective.

That's true in general, but the dep inference machinery takes care of this already. The responsibility of this extraction phase is just to say "this file depends on this module". Other code takes care of figuring out who provides the module, which is where source roots come in.

Reply:
Yea, that's what I mean. If the extraction phase says "we have a dependency on module `foo.bar`" but we only have `some.foo.bar` (due to how the source roots are set up) then the extraction phase needs to provide `some.foo.bar` for it to work, right?

Reply:
The extraction has to get the module paths right, of course, but this should not require it to know about source roots, since `import` statements (or any other runtime loading/dependency mechanisms) don't know about them.

Reply:
Agreed. That's _exactly_ what I was trying to say with
>  I guess it doesn't need to know about source roots or file paths, but the dependency module won't work if it isn't complete from a source root perspective.

Reply:
why is human language so hard to use for communication? :P